### PR TITLE
Added model to audit hooks execution

### PR DESCRIPTION
**Pull request by mariajgrimaldi**
This PR adds a model to save the result of the execution of an action

For now, the information that is saved in the DB is the information about the execution of the action. What other logged information should we save?

### CODE REVIEW

File: eox_hooks/models.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    status = models.PositiveIntegerField(
+        choices=Status.choices(), default=Status.SUCCESS, editable=False,
+    )
+    executed_at = models.DateTimeField(auto_now_add=True, editable=False)
+    hook_name = models.CharField(max_length=150)
+    task_name = models.CharField(max_length=150)
+    tenant_domain = models.CharField(max_length=150)
+    task_duration = models.FloatField(null=True)
+
+    objects = models.Manager()
```
Comment (andrey-canon) [inline eox_hooks/models.py:30]: this is the default value , isn't it ?

File: eox_hooks/tests/test_models.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+"""
+import time
+
+from django.test import TestCase
+
+from eox_hooks.constants import Status
+from eox_hooks.models import HookExecutionAudit
+
+
+class TestAuditModel(TestCase):
```
Comment (andrey-canon) [inline eox_hooks/tests/test_models.py:14]: since you are not changing the standard django behavior, I reckon these tests are not necessary, maybe test the __str__ method

File: eox_hooks/models.py (hunk 1 above)
Comment (mariajgrimaldi) [reply eox_hooks/models.py:30]: Yes! but pylint throws an error when using objects.create.
E: 19,22: Class 'HookExecutionAudit' has no 'objects' member (no-member)
I fixed it adding the default

File: eox_hooks/models.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    )
+
+    status = models.PositiveIntegerField(
+        choices=Status.choices(), default=Status.SUCCESS, editable=False,
+    )
+    executed_at = models.DateTimeField(auto_now_add=True, editable=False)
+    hook_name = models.CharField(max_length=150)
+    task_name = models.CharField(max_length=150)
+    tenant_domain = models.CharField(max_length=150)
+    task_duration = models.FloatField(null=True)
```
[the remark this answers is not in the document]
Comment (mariajgrimaldi) [reply eox_hooks/models.py:28]: I'm measuring the execution time using the built in module _time_ via time.time() that returns the seconds and a decimal part with microseconds. FloatField is not a better option? Considering how we measure time

File: eox_hooks/tests/test_models.py (hunk 2 above)
Comment (mariajgrimaldi) [reply eox_hooks/tests/test_models.py:14]: okok that's better

Comment (mariajgrimaldi) [reply eox_hooks/tests/test_models.py:14]: I'm going to leave the first test so I can check that I can save task_duration using time.time()

Comment (andrey-canon) [reply eox_hooks/tests/test_models.py:14]: we already know that floatfield works with float values and time() returns a float ....

File: eox_hooks/models.py (hunk 1 above)
Comment (andrey-canon) [reply eox_hooks/models.py:30]: Try to suppress the warning instead of add this

File: eox_hooks/tests/test_models.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        """Function that checks the correct behavior of the str model method."""
+        audit_entry = HookExecutionAudit.objects.create(
+            status=Status.SUCCESS,
+            hook_name="hook_name",
+            task_name="task_name",
+            tenant_domain="tenant_domain",
+            task_duration=time.time(),
+        )
+
+        expected_output = "Execution {key}: Task '{task_name}' executed at {time} with status {status}." \
```
Comment (andrey-canon) [inline eox_hooks/tests/test_models.py:39]: unreadable for my :(

File: eox_hooks/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrey-canon) [reply eox_hooks/models.py:28]: you can use the float value to create a timedelta

File: eox_hooks/actions_handler.py [hunk 5]
```diff
@@ -51,3 +55,18 @@ def action_lookup(module_name, action):
 
     log.warning(message)
     return custom_action_mock
+
+
+def audit_action_execution(status, sender, action, execution_time=None):
```
Comment (andrey-canon) [inline eox_hooks/actions_handler.py:60]: ???

Comment (mariajgrimaldi) [reply eox_hooks/actions_handler.py]: This function is now used in action_handler

File: eox_hooks/actions_handler.py [hunk 6]
```diff
@@ -37,17 +41,28 @@ def action_lookup(module_name, action):
         module = import_module(module_name)
         return getattr(module, action)
     except ImportError:
-        message = "The module {} with the action {} does not exist. A default action will be used."\
-                  .format(
-                      module_name,
-                      action,
-                    )
+        message = "The module {} with the action {} does not exist. A default action will be used.".format(module_name,
```
Comment (andrey-canon) [inline eox_hooks/actions_handler.py:44]: ugly
image

File: eox_hooks/tests/test_actions_handler.py [hunk 7]
```diff
@@ -32,20 +32,21 @@ def test_executing_custom_action(self, action_lookup_mock):
 
         action_mock.assert_called_once()
 
+    @patch("eox_hooks.models.HookExecutionAudit.objects.create")
     @patch("eox_hooks.actions_handler.action_lookup")
-    def test_action_fail_silently(self, action_lookup_mock):
+    def test_action_fail_silently(self, action_lookup_mock, audit_model):
```
[the remark this answers is not in the document]
Comment (mariajgrimaldi) [reply eox_hooks/tests/test_actions_handler.py]: Ok I mocked audit_action_execution and added a new test class for it, there I check the log

File: eox_hooks/actions_handler.py (hunk 6 above)
Comment (mariajgrimaldi) [reply eox_hooks/actions_handler.py]: I fixed it :100:

File: eox_hooks/admin.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
+"""
+Admin class.
+"""
+from django.contrib import admin
+
+from eox_hooks.models import HookExecutionAudit
+
+
+class HookExecutionAuditAdmin(admin.ModelAdmin):
+    """Audit admin."""
```
Comment (felipemontoya) [inline eox_hooks/admin.py:10]: Is there a way to make this admin class so that not even a super user can tamper with this DATA?

File: eox_hooks/actions_handler.py [hunk 9]
```diff
@@ -18,9 +22,9 @@ def action_handler(trigget_event, configuration, **kwargs):
 
     try:
         action(**kwargs)
-        log.info("The action {} with triggered by {} ended successfully.".format(action, trigget_event))
+        audit_action_execution(Status.SUCCESS, trigget_event, action)
     except Exception as exception:  # pylint: disable=broad-except
-        log.error("The action {} with triggered by {} failed.".format(action, trigget_event))
+        audit_action_execution(Status.FAIL, trigget_event, action)
```
Comment (felipemontoya) [inline eox_hooks/actions_handler.py:27]: do we want to write to the audit model if an Exception was raised, and the message from said exception?

File: eox_hooks/actions_handler.py [hunk 10]
```diff
@@ -37,17 +41,32 @@ def action_lookup(module_name, action):
...
-                      module.__name__,
-                    )
+        message = "The action {} does not exist in the module {}. A default action will be used.".format(
+            action,
+            module_name,
+        )
 
-    log.warning(message)
+    log.warning("EOX_HOOKS | {}".format(message))
     return custom_action_mock
```
Comment (felipemontoya) [inline eox_hooks/actions_handler.py:55]: Are we still returning a custom_action_mock? I thought this had been removed since we dont want to run anything if there is no action configured. Perhaps that belongs in a different PR

File: eox_hooks/models.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+    status = models.PositiveIntegerField(
+        choices=Status.choices(), default=Status.SUCCESS,
+    )
+    executed_at = models.DateTimeField(auto_now_add=True, editable=False)
+    trigger_event = models.CharField(max_length=150)
+    action_name = models.CharField(max_length=150)
+    tenant_domain = models.CharField(max_length=150)
+    action_duration = models.DurationField(null=True, blank=True)
+
```
Comment (felipemontoya) [inline eox_hooks/models.py:29]: Is it possible to add a text field here where we can write the whole execution log? with timestamps and tracebacks and all that

File: eox_hooks/actions_handler.py (hunk 10 above)
Comment (mariajgrimaldi) [reply eox_hooks/actions_handler.py:66]: I made this PR for that! https://github.com/eduNEXT/eox-hooks/pull/6

File: eox_hooks/admin.py (hunk 8 above)
Comment (mariajgrimaldi) [reply eox_hooks/admin.py:10]: I think we can override `has_add_permission` and `has_delete_permission` so they always return False

File: eox_hooks/models.py (hunk 11 above)
Comment (mariajgrimaldi) [reply eox_hooks/models.py:30]: Well, I added the text field. I guess we could extract that info from an exception

File: eox_hooks/actions_handler.py (hunk 9 above)
Comment (mariajgrimaldi) [reply eox_hooks/actions_handler.py]: That would be helpful and it would give more context!

**Comment by andrey-canon** [comment]
rebase
